langchain4j-vertex-ai/src/main/java/dev/langchain4j/model/vertexai/VertexAiImagenImageModel.java

```diff
@@ -0,0 +1,403 @@
+package dev.langchain4j.model.vertexai;
+
+import com.google.cloud.aiplatform.v1.EndpointName;
+import com.google.cloud.aiplatform.v1.PredictResponse;
+import com.google.cloud.aiplatform.v1.PredictionServiceClient;
+import com.google.cloud.aiplatform.v1.PredictionServiceSettings;
+import com.google.protobuf.Value;
+import com.google.protobuf.util.JsonFormat;
+import dev.langchain4j.data.image.Image;
+import dev.langchain4j.model.image.ImageModel;
+import dev.langchain4j.model.output.Response;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.google.protobuf.Value.newBuilder;
+import static dev.langchain4j.internal.Json.toJson;
+import static dev.langchain4j.internal.RetryUtils.withRetry;
+import static dev.langchain4j.internal.ValidationUtils.ensureNotBlank;
+import static java.util.Collections.singletonList;
+
+/**
+ * Image model for the Google Cloud Vertex AI Imagen image generation models.
+ * Supports both versions of Imagen v1 and v2, respectively identified by the model names
+ * <code>imagegeneration@002</code> and <code>imagegeneration@005</code>.
+ */
+public class VertexAiImagenImageModel implements ImageModel {
+
+    private final Long seed;
+    private final String endpoint;
+    private final EndpointName endpointName;
+    private final String language;
+    private final Image image;
+    private final Image mask;
+    private final Integer guidanceScale;
+    private final String negativePrompt;
+    private final ImageStyle sampleImageStyle;
+    private final Integer sampleImageSize;
+    private final int maxRetries;
+    private final Boolean withPersisting;
+    private Path tempDirectory;
+
+    /**
+     * Image style can be specified for imagen@002.
+     * For imagen@005 (Imagen v2), specify the style in the prompt instead.
+     */
+    public enum ImageStyle {
+        photograph,
+        digital_art,
+        landscape,
+        sketch,
+        watercolor,
+        cyberpunk,
+        pop_art
+    }
+
+    /**
+     * Constructor of the Imagen image generation model.
+     *
+     * @param endpoint         the base URL of the API (eg. <code>https://us-central1-aiplatform.googleapis.com/v1/</code>)
+     * @param project          the Google Cloud Project ID
+     * @param location         the cloud region (eg. <code>us-central1</code>)
+     * @param publisher        the publisher of the model (<code>google</code> for Imagen)
+     * @param modelName        the name of the image model (<code>imagegeneration@002</code> or <code>imagegeneration@005</code>)
+     * @param seed             a fixed random seed number between 0 and 2^32-1
+     * @param language         the spoken language used for the prompt
+     * @param image            an image to edit
```

Comment:
Maybe we need a separate method in `ImageModel` for image edition, like `Response<Image> edit(Image image, Image mask, String prompt);`?

Setting image and mask in the constructor feels hacky

Reply:
I didn't want to change `ImageModel` since it's also used by OpenAI and AzureOpenAI, but I also felt it was not ideal to put those parameters in the constructor, instead of per-generation.
I'll add a new default method in the `ImageModel`.

Reply:
Actually, maybe several methods...

It's possible to generate an image from:
* a single text prompt
* an image with a text prompt (it updates the image but without a mask, as if the mask was the full picture)
* an image with a mask and a text prompt (to update only a portion of an image)

But then, those 3 methods also work for generating `n` images. So that would be 2 * 3 new methods.
What do you think?

Reply:
Also I'm wondering if we should call them `generate()` or if we should use a different method name? (`edit`)

Reply:
It's actually a little trickier.
Image editing can have a scale guidance, ie. how much the image should be edited within the mask.
An upscale operation is also an image editing operation, and requires an image size.
Not all image model will support all these things, and other image models may also support other things that Imagen doesn't support.
I'm not yet sure how to best approach this.

Reply:
Thinking of creating an `EditOperation` and an `UpscaleOperation` classes, but what are the commonalities between image models?

Reply:
@glaforge I would keep it simple as of now and just add 1 or 2 methods:
`Response<Image> edit(Image image, String prompt);`
`Response<Image> edit(Image image, Image mask, String prompt);`
This should work for both VertexAI and OpenAI.
We can always add more methods if required (with `n`).

Scale guidance and other parameters for image edition can be set in the ctor for now (as we do for temperature, maxTokens, etc).
But the most variable part will be the image itself, right? This can be an argument to the `edit` method to make it flexible.

Reply:
Thanks for the guidance!

I was less worried about the `n` case, as in my mind, the `1` case is just a special case of `n` (at least that's the case with Imagen) 😄  So I was just thinking of redirecting `1` to `n` in my implementation. But I can indeed get rid of the `n` case.

For `edit`, I'll go as suggested and keep the other parameters in the constructor.
But somehow, I still feel a bit weird that we can't override temperature and max tokens on a per call basis and have to go through the constructor for this. But it's not really costly anyway to create a new instance, so I guess that's fine.

There's also the case of the _upscale_ operation. There's an image size parameter that is passed by constructor, and a mode variable, but that's only set transparently when we specify an image size. I guess I can do it as an edit operation, and have it in the constructor, just like scale guidance and others.

There's one last thing as well, which is the negative prompt. It seems lots of image models allow this (at least in their UIs, like StableDiffusion, Leonardo.AI, and indeed Imagen). Currently it's also in the constructor, but I feel like somehow this is quite integral to what an image model should support, so maybe later on we might have to support both a prompt String and a negative prompt String.

Reply:
Regarding `n`, in my understanding in 95% cases `n=1`, so from the API perspective, it will be annoying for the user to always provide `n=1` and then do `.get(0)` all the time. If my current understanding is wrong and users will ask for a method with `n`, we can always add it, right?

Regarding other parameters, I agree that current API forces user to setup all parameters during model construction and does not allow to override them during usage. The idea was that these models will be mostly used in the automated way, so once the developer found fitting parameters for a specific task, he creates a model with concrete parameters and then the only variable thing during usage is the user input (text, images, etc).
So far I did not hear complaints in this regards. But I agree, we should think about making this more flexible, but also without overloading API with dozens of methods.

Reply:
I think I'm good now.
Can you please have another look and tell me what you think about it?

Reply:
I had forgotten to put the @Disabled annotation back and also point to your GCP project instead of mine. Now it's correct.